Modules/socketmodule.c

```diff
@@ -3069,37 +3088,127 @@ sock_settimeout(PySocketSockObject *s, PyObject *arg)
+}
+
+static int
+internal_settimeout(PySocketSockObject *s, _PyTime_t timeout) {
+    int block;
 
     s->sock_timeout = timeout;
 
-    int block = timeout < 0;
-    /* Blocking mode for a Python socket object means that operations
-       like :meth:`recv` or :meth:`sendall` will block the execution of
-       the current thread until they are complete or aborted with a
-       `TimeoutError` or `socket.error` errors.  When timeout is `None`,
-       the underlying FD is in a blocking mode.  When timeout is a positive
-       number, the FD is in a non-blocking mode, and socket ops are
-       implemented with a `select()` call.
-
-       When timeout is 0.0, the FD is in a non-blocking mode.
-
-       This table summarizes all states in which the socket object and
-       its underlying FD can be:
-
-       ==================== ===================== ==============
-        `gettimeout()`       `getblocking()`       FD
-       ==================== ===================== ==============
-        ``None``             ``True``              blocking
-        ``0.0``              ``False``             non-blocking
-        ``> 0``              ``True``              non-blocking
-    */
+    if (s->sock_family == AF_INET || s->sock_family == AF_INET6) {
+        if (timeout == 0) {
+            block = 0;
+        } else if (timeout < 0) {
+            block = 1;
+
+            #ifdef MS_WINDOWS
+            DWORD zero = 0;
+            if (setsockopt(s->sock_fd, SOL_SOCKET, SO_SNDTIMEO, (char *)&zero,
+                           sizeof(DWORD)) != 0)
+            #else
+            struct timeval zero = {
+                .tv_sec = 0,
+                .tv_usec = 0,
+            };
+            if (setsockopt(s->sock_fd, SOL_SOCKET, SO_SNDTIMEO, (char *)&zero,
+                           sizeof(struct timeval)) != 0)
+            #endif
+            {
+                // EINVAL means emote closed the socket fd or shutdown has been
+                // called.
+                if (!CHECK_ERRNO(EINVAL)) {
+                    set_error();
+                    return -1;
+                }
+            }
+            #ifdef MS_WINDOWS
+            if (setsockopt(s->sock_fd, SOL_SOCKET, SO_RCVTIMEO, (char *)&zero,
+                           sizeof(int)) != 0)
+            #else
+            if (setsockopt(s->sock_fd, SOL_SOCKET, SO_RCVTIMEO, (char *)&zero,
+                           sizeof(struct timeval)) != 0)
+            #endif
+            {
+                // EINVAL means emote closed the socket fd or shutdown has been
+                // called.
+                if (!CHECK_ERRNO(EINVAL)) {
+                    set_error();
+                    return -1;
+                }
+            }
+        } else {
+            block = 1;
+
+            #ifdef MS_WINDOWS
+            DWORD timeout_ms = timeout / 1000000;
+            if (setsockopt(s->sock_fd, SOL_SOCKET, SO_SNDTIMEO,
+                           (char *)&timeout_ms, sizeof(DWORD)) != 0)
```

Comment:
If we rename `timeout_ms` to `timeout_tv`, then a few of these ifdefs can combine into:

```suggestion
            if (setsockopt(s->sock_fd, SOL_SOCKET, SO_SNDTIMEO,
                           (char *)&timeout_tv, sizeof(timeout_tv)) != 0)
```

Reply:
Got it, updated!